kafka/consumer_test.go

```diff
@@ -421,26 +438,44 @@ func TestConsumerGracefulShutdown(t *testing.T) {
 				return nil
 			}),
 		})
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		var consumerErr error
+		consumerDone := make(chan struct{})
+		go func() {
+			defer close(consumerDone)
+			consumerErr = consumer.Run(ctx)
+		}()
 
 		b, err := codec.Encode(model.APMEvent{Transaction: &model.Transaction{ID: "1"}})
 		require.NoError(t, err)
-		ctx, cancel := context.WithCancel(context.Background())
-		defer cancel()
 		for i := 0; i < records; i++ {
 			produceRecord(ctx, t, client, &kgo.Record{Topic: "topic", Value: b})
 		}
-
-		go func() { consumer.Run(ctx) }()
 		select {
 		case process <- struct{}{}:
 			close(process) // Allow records to be processed
-			cancel()       // Stop the consumer.
 		case <-time.After(time.Second):
 			t.Fatal("timed out waiting for consumer to process event")
 		}
-		assert.Eventually(t, func() bool {
-			return processed.Load() == int32(records) && errored.Load() == 0
-		}, 6*time.Second, time.Millisecond)
+		assert.NoError(t, consumer.Close())
+
+		select {
+		case <-consumerDone:
+		case <-time.After(time.Second):
+			t.Fatal("timed out waiting for consumer to exit")
+		}
+
+		// Assert the consumer error.
+		switch dt {
+		case apmqueue.AtMostOnceDeliveryType:
+			if consumerErr != nil {
+				assert.Equal(t, ErrCommitFailed, consumerErr)
+			}
```

Comment:
Should this error be consistent/deterministic?

Reply:
correct, fixed.